### ISSUE TITLE
Bottom up with ID stops too early

### ISSUE THREAD

**Issue by catubc** [murthylab/sleap#540]
Hello
We are working on training some NNs using the bottom up with id branch. 
The training seems to stop quit early and I was not sure if this was correct, or if there's a bug somewhere.
Any advice is appreciated.

Thanks so much,
catubc

INFO:sleap.nn.training:Creating tf.data.Datasets for training data generation...
INFO:sleap.nn.training:Finished creating training datasets. [48.1s]
INFO:sleap.nn.training:Starting training loop...
Epoch 1/300
WARNING:tensorflow:Callbacks method `on_train_batch_end` is slow compared to the batch time (batch time: 0.0364s vs `on_train_batch_end` time: 0.1418s). Check your callbacks.
200/200 - 78s - loss: 9.5870e-04 - MultiInstanceConfmapsHead_loss: 4.1852e-04 - ClassMapsHead_loss: 0.0540 - val_loss: 5.1159e-04 - val_MultiInstanceConfmapsHead_loss: 3.7289e-04 - val_ClassMapsHead_loss: 0.0139
Epoch 2/300
200/200 - 78s - loss: 5.0408e-04 - MultiInstanceConfmapsHead_loss: 3.6929e-04 - ClassMapsHead_loss: 0.0135 - val_loss: 4.9864e-04 - val_MultiInstanceConfmapsHead_loss: 3.6551e-04 - val_ClassMapsHead_loss: 0.0133
Epoch 3/300
200/200 - 77s - loss: 5.0186e-04 - MultiInstanceConfmapsHead_loss: 3.6856e-04 - ClassMapsHead_loss: 0.0133 - val_loss: 5.1069e-04 - val_MultiInstanceConfmapsHead_loss: 3.7561e-04 - val_ClassMapsHead_loss: 0.0135
Epoch 4/300
200/200 - 80s - loss: 4.9607e-04 - MultiInstanceConfmapsHead_loss: 3.6514e-04 - ClassMapsHead_loss: 0.0131 - val_loss: 4.9562e-04 - val_MultiInstanceConfmapsHead_loss: 3.6367e-04 - val_ClassMapsHead_loss: 0.0132
Epoch 5/300
200/200 - 72s - loss: 4.9312e-04 - MultiInstanceConfmapsHead_loss: 3.6378e-04 - ClassMapsHead_loss: 0.0129 - val_loss: 5.0538e-04 - val_MultiInstanceConfmapsHead_loss: 3.7289e-04 - val_ClassMapsHead_loss: 0.0132
Epoch 6/300
200/200 - 73s - loss: 4.9094e-04 - MultiInstanceConfmapsHead_loss: 3.6332e-04 - ClassMapsHead_loss: 0.0128 - val_loss: 5.0110e-04 - val_MultiInstanceConfmapsHead_loss: 3.6992e-04 - val_ClassMapsHead_loss: 0.0131
Epoch 7/300
200/200 - 75s - loss: 4.9254e-04 - MultiInstanceConfmapsHead_loss: 3.6391e-04 - ClassMapsHead_loss: 0.0129 - val_loss: 4.9515e-04 - val_MultiInstanceConfmapsHead_loss: 3.6554e-04 - val_ClassMapsHead_loss: 0.0130
Epoch 8/300
200/200 - 74s - loss: 4.9082e-04 - MultiInstanceConfmapsHead_loss: 3.6326e-04 - ClassMapsHead_loss: 0.0128 - val_loss: 5.1098e-04 - val_MultiInstanceConfmapsHead_loss: 3.7925e-04 - val_ClassMapsHead_loss: 0.0132
Epoch 9/300
200/200 - 74s - loss: 4.9229e-04 - MultiInstanceConfmapsHead_loss: 3.6660e-04 - ClassMapsHead_loss: 0.0126 - val_loss: 4.9264e-04 - val_MultiInstanceConfmapsHead_loss: 3.6670e-04 - val_ClassMapsHead_loss: 0.0126
Epoch 10/300
200/200 - 74s - loss: 4.8720e-04 - MultiInstanceConfmapsHead_loss: 3.6357e-04 - ClassMapsHead_loss: 0.0124 - val_loss: 5.0301e-04 - val_MultiInstanceConfmapsHead_loss: 3.7585e-04 - val_ClassMapsHead_loss: 0.0127
Epoch 11/300
200/200 - 74s - loss: 4.8462e-04 - MultiInstanceConfmapsHead_loss: 3.6261e-04 - ClassMapsHead_loss: 0.0122 - val_loss: 4.8381e-04 - val_MultiInstanceConfmapsHead_loss: 3.6220e-04 - val_ClassMapsHead_loss: 0.0122
Epoch 12/300
200/200 - 72s - loss: 4.9674e-04 - MultiInstanceConfmapsHead_loss: 3.7012e-04 - ClassMapsHead_loss: 0.0127 - val_loss: 6.2993e-04 - val_MultiInstanceConfmapsHead_loss: 4.6352e-04 - val_ClassMapsHead_loss: 0.0166
Epoch 13/300
200/200 - 72s - loss: 4.9386e-04 - MultiInstanceConfmapsHead_loss: 3.6252e-04 - ClassMapsHead_loss: 0.0131 - val_loss: 5.0943e-04 - val_MultiInstanceConfmapsHead_loss: 3.7431e-04 - val_ClassMapsHead_loss: 0.0135
Epoch 14/300
200/200 - 73s - loss: 4.9270e-04 - MultiInstanceConfmapsHead_loss: 3.6147e-04 - ClassMapsHead_loss: 0.0131 - val_loss: 5.0579e-04 - val_MultiInstanceConfmapsHead_loss: 3.7051e-04 - val_ClassMapsHead_loss: 0.0135
Epoch 15/300
200/200 - 72s - loss: 4.9863e-04 - MultiInstanceConfmapsHead_loss: 3.6692e-04 - ClassMapsHead_loss: 0.0132 - val_loss: 5.1059e-04 - val_MultiInstanceConfmapsHead_loss: 3.7479e-04 - val_ClassMapsHead_loss: 0.0136
Epoch 16/300
200/200 - 71s - loss: 4.9385e-04 - MultiInstanceConfmapsHead_loss: 3.6230e-04 - ClassMapsHead_loss: 0.0132 - val_loss: 5.0100e-04 - val_MultiInstanceConfmapsHead_loss: 3.6821e-04 - val_ClassMapsHead_loss: 0.0133
Epoch 17/300
200/200 - 71s - loss: 4.9723e-04 - MultiInstanceConfmapsHead_loss: 3.6557e-04 - ClassMapsHead_loss: 0.0132 - val_loss: 5.0600e-04 - val_MultiInstanceConfmapsHead_loss: 3.7087e-04 - val_ClassMapsHead_loss: 0.0135
Epoch 18/300
200/200 - 73s - loss: 4.9602e-04 - MultiInstanceConfmapsHead_loss: 3.6436e-04 - ClassMapsHead_loss: 0.0132 - val_loss: 5.0565e-04 - val_MultiInstanceConfmapsHead_loss: 3.7009e-04 - val_ClassMapsHead_loss: 0.0136
Epoch 19/300
200/200 - 70s - loss: 4.9842e-04 - MultiInstanceConfmapsHead_loss: 3.6612e-04 - ClassMapsHead_loss: 0.0132 - val_loss: 4.9701e-04 - val_MultiInstanceConfmapsHead_loss: 3.6315e-04 - val_ClassMapsHead_loss: 0.0134
Epoch 20/300
200/200 - 75s - loss: 4.9932e-04 - MultiInstanceConfmapsHead_loss: 3.6707e-04 - ClassMapsHead_loss: 0.0132 - val_loss: 4.9850e-04 - val_MultiInstanceConfmapsHead_loss: 3.6464e-04 - val_ClassMapsHead_loss: 0.0134
Epoch 21/300

Epoch 00021: ReduceLROnPlateau reducing learning rate to 9.999999747378752e-06.
200/200 - 69s - loss: 4.9280e-04 - MultiInstanceConfmapsHead_loss: 3.6140e-04 - ClassMapsHead_loss: 0.0131 - val_loss: 5.0882e-04 - val_MultiInstanceConfmapsHead_loss: 3.7333e-04 - val_ClassMapsHead_loss: 0.0135
Epoch 22/300
200/200 - 73s - loss: 4.9361e-04 - MultiInstanceConfmapsHead_loss: 3.6307e-04 - ClassMapsHead_loss: 0.0131 - val_loss: 5.1373e-04 - val_MultiInstanceConfmapsHead_loss: 3.7680e-04 - val_ClassMapsHead_loss: 0.0137
Epoch 23/300
200/200 - 70s - loss: 4.9412e-04 - MultiInstanceConfmapsHead_loss: 3.6289e-04 - ClassMapsHead_loss: 0.0131 - val_loss: 4.9289e-04 - val_MultiInstanceConfmapsHead_loss: 3.6183e-04 - val_ClassMapsHead_loss: 0.0131
Epoch 24/300
200/200 - 74s - loss: 4.9320e-04 - MultiInstanceConfmapsHead_loss: 3.6229e-04 - ClassMapsHead_loss: 0.0131 - val_loss: 5.0364e-04 - val_MultiInstanceConfmapsHead_loss: 3.6956e-04 - val_ClassMapsHead_loss: 0.0134
Epoch 25/300
200/200 - 69s - loss: 4.9483e-04 - MultiInstanceConfmapsHead_loss: 3.6409e-04 - ClassMapsHead_loss: 0.0131 - val_loss: 5.1279e-04 - val_MultiInstanceConfmapsHead_loss: 3.7654e-04 - val_ClassMapsHead_loss: 0.0136
Epoch 26/300
200/200 - 74s - loss: 4.9342e-04 - MultiInstanceConfmapsHead_loss: 3.6236e-04 - ClassMapsHead_loss: 0.0131 - val_loss: 5.0823e-04 - val_MultiInstanceConfmapsHead_loss: 3.7204e-04 - val_ClassMapsHead_loss: 0.0136
Epoch 27/300
200/200 - 71s - loss: 4.9813e-04 - MultiInstanceConfmapsHead_loss: 3.6621e-04 - ClassMapsHead_loss: 0.0132 - val_loss: 5.0305e-04 - val_MultiInstanceConfmapsHead_loss: 3.6878e-04 - val_ClassMapsHead_loss: 0.0134
Epoch 28/300
200/200 - 70s - loss: 4.9463e-04 - MultiInstanceConfmapsHead_loss: 3.6331e-04 - ClassMapsHead_loss: 0.0131 - val_loss: 5.0613e-04 - val_MultiInstanceConfmapsHead_loss: 3.7070e-04 - val_ClassMapsHead_loss: 0.0135
Epoch 29/300
200/200 - 72s - loss: 4.9602e-04 - MultiInstanceConfmapsHead_loss: 3.6435e-04 - ClassMapsHead_loss: 0.0132 - val_loss: 5.0359e-04 - val_MultiInstanceConfmapsHead_loss: 3.6933e-04 - val_ClassMapsHead_loss: 0.0134
Epoch 30/300
200/200 - 71s - loss: 4.9560e-04 - MultiInstanceConfmapsHead_loss: 3.6413e-04 - ClassMapsHead_loss: 0.0131 - val_loss: 5.0984e-04 - val_MultiInstanceConfmapsHead_loss: 3.7424e-04 - val_ClassMapsHead_loss: 0.0136
Epoch 31/300
200/200 - 72s - loss: 4.9546e-04 - MultiInstanceConfmapsHead_loss: 3.6419e-04 - ClassMapsHead_loss: 0.0131 - val_loss: 5.0847e-04 - val_MultiInstanceConfmapsHead_loss: 3.7299e-04 - val_ClassMapsHead_loss: 0.0135
Epoch 00031: early stopping
INFO:sleap.nn.training:Finished training loop. [38.0 min]
INFO:sleap.nn.training:Deleting visualization directory: /content/drive/MyDrive/data/labels/mar_28/multiclass_bottomup/viz
INFO:sleap.nn.training:Saving evaluation metrics to model folder...
WARNING:sleap.nn.evals:Failed to compute metrics.
INFO:sleap.nn.evals:Saved predictions: /content/drive/MyDrive/data/labels/mar_28/multiclass_bottomup/labels_pr.train.slp

/usr/local/lib/python3.7/dist-packages/sleap/nn/evals.py:505: RuntimeWarning: Mean of empty slice
  "dist.avg": np.nanmean(dists),
/usr/local/lib/python3.7/dist-packages/sleap/nn/evals.py:538: RuntimeWarning: Mean of empty slice.
  mPCK = mPCK_parts.mean()
/usr/local/lib/python3.7/dist-packages/numpy/core/_methods.py:161: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
/usr/local/lib/python3.7/dist-packages/sleap/nn/evals.py:632: RuntimeWarning: Mean of empty slice.
  pair_pck = metrics["pck.pcks"].mean(axis=-1).mean(axis=-1)
/usr/local/lib/python3.7/dist-packages/numpy/core/_methods.py:154: RuntimeWarning: invalid value encountered in true_divide
  ret, rcount, out=ret, casting='unsafe', subok=False)
/usr/local/lib/python3.7/dist-packages/sleap/nn/evals.py:634: RuntimeWarning: Mean of empty slice.
  metrics["oks.mOKS"] = pair_oks.mean()

WARNING:sleap.nn.evals:Failed to compute metrics.
INFO:sleap.nn.evals:Saved predictions: /content/drive/MyDrive/data/labels/mar_28/multiclass_bottomup/labels_pr.val.slp

/usr/local/lib/python3.7/dist-packages/sleap/nn/evals.py:505: RuntimeWarning: Mean of empty slice
  "dist.avg": np.nanmean(dists),
/usr/local/lib/python3.7/dist-packages/sleap/nn/evals.py:538: RuntimeWarning: Mean of empty slice.
  mPCK = mPCK_parts.mean()
/usr/local/lib/python3.7/dist-packages/numpy/core/_methods.py:161: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
/usr/local/lib/python3.7/dist-packages/sleap/nn/evals.py:632: RuntimeWarning: Mean of empty slice.
  pair_pck = metrics["pck.pcks"].mean(axis=-1).mean(axis=-1)
/usr/local/lib/python3.7/dist-packages/numpy/core/_methods.py:154: RuntimeWarning: invalid value encountered in true_divide
  ret, rcount, out=ret, casting='unsafe', subok=False)
/usr/local/lib/python3.7/dist-packages/sleap/nn/evals.py:634: RuntimeWarning: Mean of empty slice.
  metrics["oks.mOKS"] = pair_oks.mean()

**Comment by talmo** [comment]
Hey @catubc,

For models with multiple loss terms (like ID), training can be a little unstable, so the default heuristics for plateau detection might not be ideal.

You can play with a few different config parameters to tune the early stopping (or disable it entirely):
```json
{
    "optimization": {
        "early_stopping": {
            "stop_training_on_plateau": true,
            "plateau_min_delta": 1e-08,
            "plateau_patience": 20
        }
    }
}
```
You can try increasing the patience, decreasing the min_delta (threshold change in validation loss), or just disable stopping on plateau.

You might also want to check the learning rate scheduler, which does something similar by reducing the learning rate when a plateau is detected:
```json
{
    "optimization": {
        "initial_learning_rate": 0.0001,
        "learning_rate_schedule": {
            "reduce_on_plateau": true,
            "reduction_factor": 0.1,
            "plateau_min_delta": 1e-08,
            "plateau_patience": 10,
            "plateau_cooldown": 3,
            "min_learning_rate": 1e-08
        }
    }
}
```
If the learning rate is reduced too much or too early there will be no more decrease in the loss, which will trigger early stopping as well.

The last alternative is to try increasing the number of batches per epoch, which will basically prolong training overall:
```json
{
    "optimization": {
        "min_batches_per_epoch": 400,
        "min_val_batches_per_epoch": 50,
    }
}
```

Here's a full config (edited by hand, so might need to fix it if you get an error when running it):

```json
{
    "model": {
        "backbone": {
            "unet": {
                "stem_stride": null,
                "max_stride": 64,
                "output_stride": 4,
                "filters": 32,
                "filters_rate": 1.5,
                "middle_block": true,
                "up_interpolate": true
            }
        },
        "heads": {
            "multi_class_bottomup": {
                "confmaps": {
                    "sigma": 2.0,
                    "output_stride": 4,
                    "loss_weight": 1.0,
                    "offset_refinement": false
                },
                "class_maps": {
                    "sigma": 3.0,
                    "output_stride": 8,
                    "loss_weight": 0.001
                }
            }
        }
    },
    "optimization": {
        "preload_data": true,
        "augmentation_config": {
            "rotate": true,
            "rotation_min_angle": -180,
            "rotation_max_angle": 180,
            "scale": false,
            "scale_min": 0.9,
            "scale_max": 1.1,
            "contrast": false,
            "contrast_min_gamma": 0.5,
            "contrast_max_gamma": 2.0
        },
        "online_shuffling": true,
        "shuffle_buffer_size": 128,
        "prefetch": true,
        "batch_size": 4,
        "min_batches_per_epoch": 200,
        "min_val_batches_per_epoch": 10,
        "epochs": 600,
        "initial_learning_rate": 0.0001,
        "learning_rate_schedule": {
            "reduce_on_plateau": true,
            "reduction_factor": 0.1,
            "plateau_min_delta": 1e-08,
            "plateau_patience": 10,
            "plateau_cooldown": 3,
            "min_learning_rate": 1e-08
        },
        "early_stopping": {
            "stop_training_on_plateau": true,
            "plateau_min_delta": 1e-08,
            "plateau_patience": 20
        }
    }
}
```


Let me know if that works!

Talmo

**Comment by catubc** [comment]
Thanks for the advice.
My preference would be to just remove early stopping.  Is there a risk of over-training or breaking training if we just train for 200-300 epochs instead? I'm not sure searching through this parameter space is easy for our group.
Thanks,
catubc

**Comment by talmo** [comment]
Yup, you can definitely just try training a fixed amount. Worst case is you get a bit of overfitting. You can look at the loss curves afterwards to diagnose (in `training_log.csv`).

**Comment by catubc** [comment]
Ok great, we'll try 300 epochs without early stopping.
I will post an update soon re: results.
